#### Видеокурс по теории вероятностей и математической статистике (20 ноября - 30 декабря)
Домашнее задание

*Урок 5 Проверка статистических гипотез. P-значения. Доверительные интервалы

In [1]:
import numpy as np
from scipy import stats


__Задача 1__

Известно, что генеральная совокупность распределена нормально с известным средним квадратическим отклонением 16. Найти ширину доверительного интервала для оценки математического ожидания `mu` с надёжностью 0.95, если выборочное среднее равно 80, а объём выборки равен 256.

In [2]:
X_ = 80 #выборочное среднее
n = 256 #объем выборки
sigma = 16 #среднее квадратичное отклонение
p = 0.95 #надежость
alpha = 1 - p

#если совокупность распределена нормально
t1 = stats.norm.ppf(alpha / 2)
t2 = stats.norm.ppf(1 - alpha / 2)

t1, t2

(-1.959963984540054, 1.959963984540054)

In [3]:
#ширина доверитаельного интеравала
interval = (X_ + t1 * sigma / np.sqrt(n), X_ + t2 * sigma / np.sqrt(n))
print(f'Доверительный интервал:{interval}')
interval_width = interval[1]-interval[0]
print(f'Ширина доверительного интервала: {interval_width}')


Доверительный интервал:(78.04003601545995, 81.95996398454005)
Ширина доверительного интервала: 3.9199279690801063


Ответ: Ширина доверительного интервала: 3.9199279690801063

__Задача 2__

Продавец утверждает, что средний вес пачки печенья составляет 200 г. Из партии извлечена выборка из 10 пачек. Вес каждой пачки составляет:
```
202, 203, 199, 197, 195, 201, 200, 204, 194, 190
```
Известно, что их веса распределены нормально. Верно ли утверждение продавца, если учитывать, что уровень значимости равен 1%? Найдите P-значение для данного теста и укажите его в качестве ответа.

In [4]:
alpha = 0.01 #уровень значимости
#веса распределены нормально
example=np.array([202, 203, 199, 197, 195, 201, 200, 204, 194, 190])
#Нулевая гипотеза Ho=200, дисперсия не известа
#используем распределение Стьюдента

$$t = \dfrac{\overline{X} - \mu}{\sigma_X / \sqrt{n}},$$
где $\sigma_X$ — несмещённая оценка среднего квадратического отклонения. В нашем случае это:
$$t = \dfrac{\overline{X} - 200}{\sigma_X / \sqrt{n}}$$

In [5]:
print(f'Выборочное среднее Х_: {example.mean()}')
print(f'СигмаХ_несмещенная оценка ср.кв.откл. : {example.std(ddof=1)}')

Выборочное среднее Х_: 198.5
СигмаХ_несмещенная оценка ср.кв.откл. : 4.453463071962462


In [6]:
def statistic(samples: np.ndarray) -> float:
    return (samples.mean() - 200) / (samples.std(ddof=1) / np.sqrt(samples.shape[0]))  

In [7]:
n = example.shape[0] #размер выборки
S = statistic(example) #статистический тест

n, S

(10, -1.0651074037450896)

In [8]:
n = example.shape[0]
t1 = stats.norm.ppf(alpha / 2)
t2 = stats.norm.ppf(1 - alpha / 2)

t1, t2

(-2.575829303548901, 2.5758293035489004)

In [9]:
# P-значение расчетным методом
p_left = stats.t.cdf(S, df=n - 1)
p_right = 1 - stats.t.cdf(S, df=n - 1)

pvalue = 2 * min(p_left, p_right)

pvalue

0.31456660352472854

In [10]:
# P-значение встроенной функцией
stats.ttest_1samp(example, popmean=200)

Ttest_1sampResult(statistic=-1.0651074037450896, pvalue=0.31456660352472854)

Ответ: 0.3145666

__Задача 3__

Дана выборка диаметров подшипников из примера 1 с занятия 5:
```
samples = [0.6603, 0.9466, 0.5968, 1.3792, 1.5481, 0.7515, 1.0681, 1.1134,
           1.2088, 1.701 , 1.0282, 1.3579, 1.0191, 1.1784, 1.1168, 1.1372,
           0.7273, 1.3958, 0.8665, 1.5112, 1.161 , 1.0232, 1.0865, 1.02  ]
```
Предполагая, что диаметры подшипников распределены нормально, проверьте гипотезу о том, что дисперсия случайной величины равна `0.0625` при уровне значимости `alpha = 0.05`. Что для этого нужно знать:
1. Альтернативная гипотеза двухсторонняя.
2. Статистика для теста: `H = (n - 1) * sample_variance / variance`, где `n` - число элементов в выборке, `sample_variance` - несмещённая оценка дисперсии, `variance` - утверждаемая нулевой гипотезой дисперсия.
3. Эта статистика в предположении верности нулевой гипотезы имеет распределение хи-квадрат с параметром `df = n - 1`. Её квантили можно найти с помощью функции `scipy.stats.chi2.ppf`.

В качестве ответа запишите 1, если нулевая гипотеза верна, и 0 иначе.

In [11]:
alpha = 0.05 #уровень значимости
sigma = 0.0625 #дисперсия случайной величины
#диаметры подшипников распределены нормально
example=np.array(
           [0.6603, 0.9466, 0.5968, 1.3792, 1.5481, 0.7515, 1.0681, 1.1134,
           1.2088, 1.701 , 1.0282, 1.3579, 1.0191, 1.1784, 1.1168, 1.1372,
           0.7273, 1.3958, 0.8665, 1.5112, 1.161 , 1.0232, 1.0865, 1.02]
            )
example

array([0.6603, 0.9466, 0.5968, 1.3792, 1.5481, 0.7515, 1.0681, 1.1134,
       1.2088, 1.701 , 1.0282, 1.3579, 1.0191, 1.1784, 1.1168, 1.1372,
       0.7273, 1.3958, 0.8665, 1.5112, 1.161 , 1.0232, 1.0865, 1.02  ])

In [12]:
n = example.shape[0] #число элементов в выборке
sample_variance = example.var(ddof=1) #несмещённая оценка дисперсии
variance = 0.0625 #утверждаемая нулевой гипотезой дисперсия.
H =(n - 1) * sample_variance / variance #Статистика для теста
H

28.720541753333343

In [13]:
ST=stats.chi2.ppf(1-alpha, df = n - 1) #квантиль статистики при распределении хи-квадрат
ST

35.17246162690806

критическая область определяется интервалом (35,17; +oo). т.о 28 < 35.17 значит нулевая гипотеза верна

Ответ: 1